File: app/lib/mapa/pmtiles.dart
// EL LECTOR DEL PAQUETE: PMTiles v3.
//
// Un paquete de mapa es **un solo fichero**. Dentro lleva una cabecera de 127
// bytes, unos directorios y las teselas pegadas unas detras de otras. Sacar la
// tesela `z/x/y` es: mirar el directorio, saltar al sitio que diga, leer tantos
// bytes.
//
// ## Por que PMTiles y no MBTiles, que es SQLite y ya lo tenemos
//
// Se miraron los dos y el argumento de MBTiles es de peso: la aplicacion ya
// lleva SQLite dentro para la base local, asi que leer una tesela seria una
// consulta y no este fichero. Se eligio PMTiles por tres cosas concretas:
//
//  1. **Un fichero y nada mas.** SQLite se trae sus companeros (`-wal`, `-shm`)
//     y quiere permiso de escritura en la carpeta hasta para leer. Aqui lo que
//     hay es un fichero que se baja una vez, se comprueba por `sha256` y no se
//     vuelve a tocar: cuanto menos se parezca a una base de datos viva, mejor.
//  2. **Se lee por rangos.** Hoy se baja entero, que es lo que pidio Jose
//     —«que se descargue el mapa... cuando cargue sólo una vez»—, pero el dia
//     que haga falta ensenar una provincia sin bajarse Cuba, el formato ya lo
//     permite y el codigo de aqui no cambia.
//  3. **Lo escriben las herramientas de siempre.** `tippecanoe --output=x.pmtiles`
//     y `planetiler` sacan esto mismo. Si un dia nuestro generador se queda
//     corto, se cambia el generador y no la aplicacion.
//
// Lo que se perdio al elegirlo: hay que escribir este lector. Son doscientas
// lineas y el escritor esta en el mismo repositorio
// (`herramientas/mapa-cuba/pmtiles.go`), con una muestra generada por el uno y
// leida por el otro en `test/mapa/muestra/`. Esa pareja es lo unico que
// comprueba de verdad un formato binario: un lector que solo se prueba contra
// ficheros que el mismo escribe no comprueba nada.

import 'dart:convert';
import 'dart:typed_data';

import 'descomprimir.dart';

/// De donde salen los bytes. Una interfaz y no un `File` para poder probar el
/// lector con un fichero en memoria y **sin tocar el disco**.
abstract interface class LeerPorRangos {
  /// [largo] bytes a partir de [desde]. Lanza si no los hay: pedir mas alla del
  /// final es un fichero corrupto, no un caso normal.
  Future<Uint8List> leer(int desde, int largo);

  Future<void> cerrar();
}

/// Bytes que ya estan en memoria. Es lo que usan las pruebas.
class RangosEnMemoria implements LeerPorRangos {
  RangosEnMemoria(this.bytes);

  final Uint8List bytes;

  @override
  Future<Uint8List> leer(int desde, int largo) async {
    if (desde < 0 || largo < 0 || desde + largo > bytes.length) {
      throw RangeError('se pidieron $largo bytes en $desde de ${bytes.length}');
    }
    return Uint8List.sublistView(bytes, desde, desde + largo);
  }

  @override
  Future<void> cerrar() async {}
}

/// Lo que el paquete dice de si mismo.
class CabeceraDelPaquete {
  const CabeceraDelPaquete({
    required this.raizDesde,
    required this.raizLargo,
    required this.metaDesde,
    required this.metaLargo,
    required this.hojasDesde,
    required this.datosDesde,
    required this.entradas,
    required this.zMin,
    required this.zMax,
    required this.minLon,
    required this.minLat,
    required this.maxLon,
    required this.maxLat,
  });

  final int raizDesde;
  final int raizLargo;
  final int metaDesde;
  final int metaLargo;
  final int hojasDesde;
  final int datosDesde;

  /// Cuantas teselas hay. Sirve para una cosa muy concreta: decirlo en la
  /// pantalla, para que «tienes el mapa» sea un numero y no una promesa.
  final int entradas;

  final int zMin;
  final int zMax;
  final double minLon;
  final double minLat;
  final double maxLon;
  final double maxLat;
}

/// Lo que puede salir mal al abrir un paquete. Con nombre y con motivo: «no se
/// pudo abrir el mapa» no le dice a nadie que hacer.
class PaqueteIlegible implements Exception {
  const PaqueteIlegible(this.motivo);

  final String motivo;

  @override
  String toString() => motivo;
}

/// EL PAQUETE ABIERTO.
class PaqueteDeTeselas {
  PaqueteDeTeselas._(this._fuente, this.cabecera);

  final LeerPorRangos _fuente;
  final CabeceraDelPaquete cabecera;

  /// Los directorios ya leidos. Un directorio hoja cubre miles de teselas, asi
  /// que recordarlo ahorra una lectura por cada tesela de la misma zona — que en
  /// un telefono, con el fichero en la tarjeta, se nota.
  final _directorios = <int, List<_Entrada>>{};

  static const _tamanoDeCabecera = 127;

  /// Abre el paquete. Lanza [PaqueteIlegible] con el motivo si no lo es.
  static Future<PaqueteDeTeselas> abrir(LeerPorRangos fuente) async {
    Uint8List crudo;
    try {
      crudo = await fuente.leer(0, _tamanoDeCabecera);
    } on Object {
      throw const PaqueteIlegible(
        'el fichero del mapa tiene menos de 127 bytes: se quedó a medio bajar',
      );
    }
    final b = ByteData.sublistView(crudo);
    if (String.fromCharCodes(crudo.sublist(0, 7)) != 'PMTiles') {
      throw const PaqueteIlegible(
        'el fichero del mapa no empieza por «PMTiles»: no es un paquete de mapa',
      );
    }
    if (crudo[7] != 3) {
      throw PaqueteIlegible(
        'el paquete es de la versión ${crudo[7]} de PMTiles y esta aplicación '
        'lee la 3',
      );
    }
    // Las dos compresiones tienen que ser gzip y las teselas MVT. Si no lo son
    // se dice: leerlo igual da un mapa de garabatos, que es peor que no tenerlo.
    if (crudo[97] != 2 || crudo[98] != 2) {
      throw const PaqueteIlegible(
        'el paquete viene comprimido de otra forma; esta aplicación lee gzip',
      );
    }
    if (crudo[99] != 1) {
      throw const PaqueteIlegible(
        'el paquete trae teselas de imagen y no vectoriales: es el formato que '
        'ocupa 6,5 GB en vez de 26 MB',
      );
    }

    double grados(int pos) => b.getInt32(pos, Endian.little) / 1e7;
    final cabecera = CabeceraDelPaquete(
      raizDesde: b.getUint64(8, Endian.little),
      raizLargo: b.getUint64(16, Endian.little),
      metaDesde: b.getUint64(24, Endian.little),
      metaLargo: b.getUint64(32, Endian.little),
      hojasDesde: b.getUint64(40, Endian.little),
      datosDesde: b.getUint64(56, Endian.little),
      entradas: b.getUint64(80, Endian.little),
      zMin: crudo[100],
      zMax: crudo[101],
      minLon: grados(102),
      minLat: grados(106),
      maxLon: grados(110),
      maxLat: grados(114),
    );
    if (cabecera.entradas == 0) {
      // UNA RESPUESTA VACÍA NO ES UNA RESPUESTA BUENA (`CLAUDE.md` §3). Un
      // paquete sin una sola tesela se abre sin un error y deja el mapa en
      // blanco para siempre.
      throw const PaqueteIlegible(
        'el paquete no trae ni una tesela: el fichero está vacío o se generó mal',
      );
    }
    return PaqueteDeTeselas._(fuente, cabecera);
  }

  Future<void> cerrar() => _fuente.cerrar();

  /// Los metadatos del paquete, incluida **la atribución de OpenStreetMap**.
  /// Viaja dentro del fichero a propósito: la licencia la exige también sin
  /// conexión, y metida aquí no se pierde aunque cambie la pantalla.
  Future<Map<String, Object?>> metadatos() async {
    if (cabecera.metaLargo == 0) return const {};
    try {
      final crudo = await _fuente.leer(cabecera.metaDesde, cabecera.metaLargo);
      final texto = utf8.decode(descomprimirGzip(crudo), allowMalformed: true);
      final leido = jsonDecode(texto);
      return leido is Map<String, Object?> ? leido : const {};
    } on Object {
      return const {};
    }
  }

  /// LA TESELA `z/x/y`, **ya descomprimida**, o `null` si el paquete no la trae.
  ///
  /// `null` es lo normal, no una avería: el paquete sólo lleva las teselas que
  /// tienen algo, y en Cuba casi toda la caja es mar.
  Future<Uint8List?> tesela(int z, int x, int y) async {
    if (z < cabecera.zMin || z > cabecera.zMax) return null;
    final buscado = idDeTesela(z, x, y);
    if (buscado == null) return null;

    var desde = cabecera.raizDesde;
    var largo = cabecera.raizLargo;
    // Tres saltos son de sobra: la especificación no encadena más de dos niveles
    // de hoja. El tope está para que un fichero corrupto que se apunte a sí
    // mismo no deje la aplicación dando vueltas para siempre.
    for (var salto = 0; salto < 3; salto++) {
      final entradas = await _directorio(desde, largo);
      final e = _buscar(entradas, buscado);
      if (e == null) return null;
      if (e.repetida != 0) {
        try {
          final crudo = await _fuente.leer(cabecera.datosDesde + e.desde, e.largo);
          return descomprimirGzip(crudo);
        } on Object {
          // Una tesela ilegible es UNA tesela, no el paquete. Se devuelve null
          // y el mapa se dibuja sin ese trozo, con el croquis debajo.
          return null;
        }
      }
      desde = cabecera.hojasDesde + e.desde;
      largo = e.largo;
    }
    return null;
  }

  Future<List<_Entrada>> _directorio(int desde, int largo) async {
    final ya = _directorios[desde];
    if (ya != null) return ya;
    final crudo = await _fuente.leer(desde, largo);
    final entradas = _leerDirectorio(descomprimirGzip(crudo));
    _directorios[desde] = entradas;
    return entradas;
  }
}

/// Una fila de directorio.
class _Entrada {
  _Entrada(this.id, this.desde, this.largo, this.repetida);

  final int id;
  final int desde;
  final int largo;

  /// `0` significa **«esto apunta a un directorio hoja»**, no «una tesela que se
  /// repite cero veces». Tratarlo como tesela devuelve el directorio comprimido
  /// donde tenía que ir un dibujo, y eso se ve como un mapa en blanco.
  final int repetida;
}

List<_Entrada> _leerDirectorio(Uint8List datos) {
  final lector = _LectorDeVarints(datos);
  final cuantas = lector.varint();
  if (cuantas <= 0 || cuantas > 1 << 24) {
    // Un número absurdo aquí es un fichero corrupto, no un fichero grande. Sin
    // este tope, reservar esa lista tumba la aplicación.
    throw PaqueteIlegible('el directorio dice tener $cuantas entradas');
  }
  final ids = List<int>.filled(cuantas, 0);
  var ultimo = 0;
  for (var i = 0; i < cuantas; i++) {
    ultimo += lector.varint();
    ids[i] = ultimo;
  }
  final repetidas = [for (var i = 0; i < cuantas; i++) lector.varint()];
  final largos = [for (var i = 0; i < cuantas; i++) lector.varint()];
  final desdes = List<int>.filled(cuantas, 0);
  for (var i = 0; i < cuantas; i++) {
    final v = lector.varint();
    // `0` no es un desplazamiento: significa «va pegada a la anterior». Es lo
    // que permite que una tira de teselas seguidas no gaste ni un byte en decir
    // dónde está cada una.
    desdes[i] = (v == 0 && i > 0)
        ? desdes[i - 1] + largos[i - 1]
        : v - 1;
  }
  return [
    for (var i = 0; i < cuantas; i++)
      _Entrada(ids[i], desdes[i], largos[i], repetidas[i]),
  ];
}

/// La búsqueda del formato: **la última entrada con `id <= buscado`**, no la que
/// tenga ese id exacto. Esa diferencia es la que permite que una entrada de hoja
/// cubra un rango entero de teselas.
_Entrada? _buscar(List<_Entrada> entradas, int buscado) {
  var bajo = 0;
  var alto = entradas.length - 1;
  while (bajo <= alto) {
    final medio = (bajo + alto) ~/ 2;
    final id = entradas[medio].id;
    if (id > buscado) {
      alto = medio - 1;
    } else if (id < buscado) {
      bajo = medio + 1;
    } else {
      return entradas[medio];
    }
  }
  if (alto < 0) return null;
  final e = entradas[alto];
  if (e.repetida == 0) return e; // una hoja: cubre lo que venga detrás
  if (buscado - e.id < e.repetida) return e;
  return null;
}

/// EL IDENTIFICADOR DE UNA TESELA: el desplazamiento de su nivel más su posición
/// en la **curva de Hilbert**.
///
/// La curva es lo que mantiene juntos en el fichero los vecinos en el mapa.
/// Escrita al revés, el fichero se abre igual, pesa igual y devuelve la tesela
/// de otro sitio — que se ve como un mapa raro, no como un error. Por eso el
/// generador de Go y esto se prueban contra los mismos números
/// (`herramientas/mapa-cuba/mapa_cuba_test.go`).
int? idDeTesela(int z, int x, int y) {
  if (z < 0 || z > 26) return null;
  final cuantas = 1 << z;
  if (x < 0 || y < 0 || x >= cuantas || y >= cuantas) return null;

  var id = ((1 << (2 * z)) - 1) ~/ 3;
  var xi = x;
  var yi = y;
  for (var s = cuantas ~/ 2; s > 0; s ~/= 2) {
    final rx = (xi & s) > 0 ? 1 : 0;
    final ry = (yi & s) > 0 ? 1 : 0;
    id += s * s * ((3 * rx) ^ ry);
    if (ry == 0) {
      if (rx == 1) {
        xi = s - 1 - xi;
        yi = s - 1 - yi;
      }
      final t = xi;
      xi = yi;
      yi = t;
    }
  }
  return id;
}

class _LectorDeVarints {
  _LectorDeVarints(this.b);

  final Uint8List b;
  int i = 0;

  int varint() {
    var valor = 0;
    var desplazamiento = 0;
    while (i < b.length) {
      final byte = b[i++];
      valor |= (byte & 0x7F) << desplazamiento;
      if (byte & 0x80 == 0) return valor;
      desplazamiento += 7;
      if (desplazamiento > 63) {
        throw const PaqueteIlegible('el directorio del paquete está corrupto');
      }
    }
    throw const PaqueteIlegible('el directorio del paquete se acaba a medias');
  }
}

File: herramientas/mapa-cuba/pmtiles.go
// EL CONTENEDOR: PMTiles v3, escrito a mano.
//
// POR QUE A MANO Y NO UNA BIBLIOTECA: el formato entero cabe en este fichero
// —una cabecera de 127 bytes y unos directorios de varints— y a cambio el
// generador no arrastra ninguna dependencia mas. La razon de peso, sin embargo,
// es otra: **el lector de Dart lo escribimos nosotros** (`app/lib/mapa/`), y
// tener el escritor delante, en el mismo repositorio, es lo unico que permite
// probar los dos contra el mismo fichero.
//
// El formato es el publico (github.com/protomaps/PMTiles, spec v3), no uno
// nuestro: un `.pmtiles` que salga de aqui se abre con `pmtiles show`, con
// protomaps.js y con cualquier herramienta de las de siempre, y al reves —un
// fichero que salga de `tippecanoe --output=x.pmtiles`— lo lee la aplicacion
// sin tocar nada. Esa es la puerta de salida si un dia este generador se queda
// corto.
//
// Lo que NO hace este escritor, a proposito:
//
//   - no escribe teselas vacias. Cuba es una isla en una caja que es casi toda
//     mar: contar el mar como tesela es lo que convierte 113.000 teselas de
//     tierra en 492.000 de caja.
//   - no repite un contenido identico. Dos teselas con los mismos bytes
//     comparten sitio (`num_tile_contents` < `num_tile_entries`).
package main

import (
	"bytes"
	"compress/gzip"
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"io"
	"math"
	"os"
	"sort"
)

// Los enumerados del formato, con nombre para no sembrar numeros sueltos.
const (
	compresionGzip = 2 // internal_compression y tile_compression
	tipoMVT        = 1 // tile_type
)

// idDeTesela traduce `z/x/y` al identificador de PMTiles: el desplazamiento del
// nivel mas la posicion en la **curva de Hilbert** de ese nivel.
//
// Por que Hilbert y no fila a fila: la curva mantiene juntos en el fichero los
// vecinos en el mapa, y asi un directorio cubre una zona en vez de una tira.
// Eso es lo que permite leer por rangos sin bajarse el fichero entero — que hoy
// no hace falta porque se baja entero, pero es la puerta que deja abierta.
func idDeTesela(z uint8, x, y uint32) uint64 {
	if z > 31 {
		panic("zoom fuera de rango")
	}
	acumulado := (uint64(1)<<(2*uint64(z)) - 1) / 3
	var rx, ry uint32
	var d uint64
	for s := uint32(1) << z / 2; s > 0; s /= 2 {
		rx, ry = 0, 0
		if x&s > 0 {
			rx = 1
		}
		if y&s > 0 {
			ry = 1
		}
		d += uint64(s) * uint64(s) * uint64((3*rx)^ry)
		// La rotacion del cuadrante, que es lo que hace que la curva sea
		// continua. Copiada letra a letra de la especificacion: cambiar el
		// orden de estas cuatro lineas da una curva que tambien "funciona" y
		// que ningun otro lector entiende.
		if ry == 0 {
			if rx == 1 {
				x = s - 1 - x
				y = s - 1 - y
			}
			x, y = y, x
		}
	}
	return acumulado + d
}

// entrada es una fila de directorio.
type entrada struct {
	ID       uint64
	Desde    uint64 // desplazamiento dentro del bloque de datos (o de las hojas)
	Largo    uint32
	Repetida uint32 // 1 en una tesela; 0 significa «esto apunta a un directorio hoja»
}

// serializarDirectorio escribe las entradas tal y como manda la especificacion:
// cuatro tiras de varints —ids delta, repeticiones, largos, desplazamientos— y
// no una tira de estructuras. Agrupar por columna es lo que hace que gzip
// encuentre algo que comprimir.
func serializarDirectorio(entradas []entrada) []byte {
	var b bytes.Buffer
	var tmp [binary.MaxVarintLen64]byte
	poner := func(v uint64) {
		n := binary.PutUvarint(tmp[:], v)
		b.Write(tmp[:n])
	}

	poner(uint64(len(entradas)))
	var ultimo uint64
	for _, e := range entradas {
		poner(e.ID - ultimo)
		ultimo = e.ID
	}
	for _, e := range entradas {
		poner(uint64(e.Repetida))
	}
	for _, e := range entradas {
		poner(uint64(e.Largo))
	}
	for i, e := range entradas {
		// `0` no es un desplazamiento: significa «va pegada a la anterior».
		// Cualquier otro valor va sumado uno, para dejar el cero libre.
		if i > 0 && e.Desde == entradas[i-1].Desde+uint64(entradas[i-1].Largo) {
			poner(0)
		} else {
			poner(e.Desde + 1)
		}
	}
	return b.Bytes()
}

func comprimir(datos []byte) []byte {
	var b bytes.Buffer
	w, _ := gzip.NewWriterLevel(&b, gzip.BestCompression)
	_, _ = w.Write(datos)
	_ = w.Close()
	return b.Bytes()
}

// El techo del directorio raiz, de la especificacion. No es un capricho: es lo
// que cabe en una peticion de rango de las que hace un navegador.
const techoDelDirectorioRaiz = 16384

// armarDirectorios parte las entradas en un directorio raiz y, si no caben, en
// hojas. Se prueba primero con todo en la raiz y se va doblando el tamano de
// hoja hasta que la raiz cabe: es el mismo procedimiento de la implementacion
// de referencia, y da ficheros que otras herramientas leen igual.
func armarDirectorios(entradas []entrada) (raiz, hojas []byte, cuantasHojas int) {
	if len(entradas) < techoDelDirectorioRaiz {
		if r := comprimir(serializarDirectorio(entradas)); len(r) <= techoDelDirectorioRaiz {
			return r, nil, 0
		}
	}
	porHoja := 4096
	for {
		var conHojas bytes.Buffer
		var deRaiz []entrada
		for i := 0; i < len(entradas); i += porHoja {
			hasta := min(i+porHoja, len(entradas))
			bloque := comprimir(serializarDirectorio(entradas[i:hasta]))
			deRaiz = append(deRaiz, entrada{
				ID:    entradas[i].ID,
				Desde: uint64(conHojas.Len()),
				Largo: uint32(len(bloque)),
				// `Repetida = 0` es LA marca de «esto es una hoja». Ponerle 1
				// haria que un lector devolviera el directorio como si fuera
				// una tesela, y eso se ve como un mapa en blanco, no como un
				// error.
				Repetida: 0,
			})
			conHojas.Write(bloque)
		}
		r := comprimir(serializarDirectorio(deRaiz))
		if len(r) <= techoDelDirectorioRaiz {
			return r, conHojas.Bytes(), len(deRaiz)
		}
		porHoja *= 2
	}
}

// Caja es el recuadro que se anuncia en la cabecera. Sale de lo que de verdad
// se metio, no de una constante: si un dia se genera solo una provincia, el
// fichero lo dice.
type Caja struct {
	MinLon, MinLat, MaxLon, MaxLat float64
}

// Escritor junta teselas y al final las vuelca. Las teselas NO se guardan en
// memoria: se escriben segun llegan a un fichero temporal de datos, y solo se
// guarda la entrada. Con 88 MB de teselas, guardarlas todas en RAM y volcarlas
// al final es la diferencia entre generar en un portatil y no generar.
type Escritor struct {
	datos       *os.File
	desde       uint64
	entradas    []entrada
	yaEstaban   map[[32]byte]entrada // dedup por contenido
	direcciones uint64               // num_addressed_tiles
}

func NuevoEscritor(temporal string) (*Escritor, error) {
	f, err := os.Create(temporal)
	if err != nil {
		return nil, err
	}
	return &Escritor{datos: f, yaEstaban: map[[32]byte]entrada{}}, nil
}

// Anadir mete la tesela `z/x/y` con su contenido YA comprimido.
func (e *Escritor) Anadir(z uint8, x, y uint32, contenido []byte) error {
	if len(contenido) == 0 {
		return nil // una tesela vacia no se escribe: ver la cabecera del fichero
	}
	e.direcciones++
	huella := sha256.Sum256(contenido)
	if vieja, hay := e.yaEstaban[huella]; hay {
		e.entradas = append(e.entradas, entrada{
			ID: idDeTesela(z, x, y), Desde: vieja.Desde, Largo: vieja.Largo, Repetida: 1,
		})
		return nil
	}
	if _, err := e.datos.Write(contenido); err != nil {
		return err
	}
	nueva := entrada{ID: idDeTesela(z, x, y), Desde: e.desde, Largo: uint32(len(contenido)), Repetida: 1}
	e.desde += uint64(len(contenido))
	e.yaEstaban[huella] = nueva
	e.entradas = append(e.entradas, nueva)
	return nil
}

// Cerrar vuelca el `.pmtiles` completo.
func (e *Escritor) Cerrar(salida string, metadatos []byte, zMin, zMax uint8, caja Caja) error {
	defer func() {
		nombre := e.datos.Name()
		_ = e.datos.Close()
		_ = os.Remove(nombre)
	}()
	if _, err := e.datos.Seek(0, io.SeekStart); err != nil {
		return err
	}

	sort.Slice(e.entradas, func(i, j int) bool { return e.entradas[i].ID < e.entradas[j].ID })
	raiz, hojas, _ := armarDirectorios(e.entradas)
	meta := comprimir(metadatos)

	f, err := os.Create(salida)
	if err != nil {
		return err
	}
	defer f.Close()

	const cabecera = 127
	desdeRaiz := uint64(cabecera)
	desdeMeta := desdeRaiz + uint64(len(raiz))
	desdeHojas := desdeMeta + uint64(len(meta))
	desdeDatos := desdeHojas + uint64(len(hojas))

	h := make([]byte, cabecera)
	copy(h, "PMTiles")
	h[7] = 3
	u64 := func(pos int, v uint64) { binary.LittleEndian.PutUint64(h[pos:], v) }
	i32 := func(pos int, v float64) {
		binary.LittleEndian.PutUint32(h[pos:], uint32(int32(math.Round(v*1e7))))
	}
	u64(8, desdeRaiz)
	u64(16, uint64(len(raiz)))
	u64(24, desdeMeta)
	u64(32, uint64(len(meta)))
	u64(40, desdeHojas)
	u64(48, uint64(len(hojas)))
	u64(56, desdeDatos)
	u64(64, e.desde)
	u64(72, e.direcciones)
	u64(80, uint64(len(e.entradas)))
	u64(88, uint64(len(e.yaEstaban)))
	// `clustered = 0`, y es la verdad, no una omision: las teselas se vuelcan en
	// el orden en que se generan —por filas— y no en el de la curva de Hilbert,
	// y ademas dos iguales comparten sitio. Poner `1` aqui le diria a un lector
	// que puede suponer desplazamientos crecientes, y es justo lo que no puede
	// suponer.
	h[96] = 0
	h[97] = compresionGzip
	h[98] = compresionGzip
	h[99] = tipoMVT
	h[100] = zMin
	h[101] = zMax
	i32(102, caja.MinLon)
	i32(106, caja.MinLat)
	i32(110, caja.MaxLon)
	i32(114, caja.MaxLat)
	h[118] = zMin
	i32(119, (caja.MinLon+caja.MaxLon)/2)
	i32(123, (caja.MinLat+caja.MaxLat)/2)

	for _, trozo := range [][]byte{h, raiz, meta, hojas} {
		if _, err := f.Write(trozo); err != nil {
			return err
		}
	}
	if _, err := io.Copy(f, e.datos); err != nil {
		return fmt.Errorf("volcando las teselas: %w", err)
	}
	return nil
}
